Revamp the guidelines around types. (#2966)
dart-lang/site-www: src/_guides/language/effective-dart/design_migrated.md

}
{% endprettify %}

### DO annotate return types on function declarations.

Dart does not generally infer the return type of a function declaration from its body,
unlike some other languages. That means you should write a type annotation for
the return type yourself.

{:.good}
<?code-excerpt "design_good.dart (annotate-return-types)"?>
{% prettify dart tag=pre+code %}
String makeGreeting(String who) {

>   return "Hello, $who!";

Author: Fixed. (I will seemingly never break this habit.)

}
{% endprettify %}

{:.bad}
<?code-excerpt "design_bad.dart (annotate-return-types)"?>
{% prettify dart tag=pre+code %}
makeGreeting(String who) {
  return "Hello, $who!";
}
{% endprettify %}

There are two cases that seem like exceptions but aren't:

*   Function *expressions* ("lambdas" or "anonymous functions") do infer a
    return type from the body. In fact, the syntax doesn't even allow a return
    type annotation. But these aren't what we consider function *declarations*.

*   The return type of an overriding method can be inferred from the method it

>     overrides. However, this is a rarely-known feature and even in this case

Author: I like "little-known".

var names = people.map((Person person) => person.name);
{% endprettify %}

If the language is able to infer the type you want for a parameter in a function
expression, then let it and don't annotate. In rare cases, the surrounding
context is not precise enough to provide a type for one or more of the
function's parameters. In those cases, you may need to annotate.

### DON'T type annotate initializing formals.

{% include linter-rule.html rule="type_init_formals" %}

If a constructor parameter is using `this.` to initialize a field, then the type
of the parameter is inferred to have the same type as the field.

{:.good}
<?code-excerpt "design_good.dart (dont-type-init-formals)"?>
{% prettify dart tag=pre+code %}
class Point {
  double x, y;
  Point(this.x, this.y);
}
{% endprettify %}

{:.bad}
<?code-excerpt "design_bad.dart (dont-type-init-formals)"?>
{% prettify dart tag=pre+code %}
class Point {
  double x, y;
  Point(double this.x, double this.y);
}
{% endprettify %}

### DO write type arguments on generic invocations that aren't inferred.

Dart is pretty smart about inferring type arguments in generic invocations. It

> will look at the expected type where the expression occurs and the types of

Reviewer: ```suggestion
looks at the expected type where the expression occurs and the types of
```

We try to stick to present tense: https://developers.google.com/style/tense

values being passed to the invocation. However, sometimes those aren't enough to
fully determine a type argument. In that case, write the entire type argument
list explicitly.

{:.good}
<?code-excerpt "design_good.dart (non-inferred-type-args)"?>
{% prettify dart tag=pre+code %}
var playerScores = <String, int>{};
final events = StreamController<Event>();
{% endprettify %}

{:.bad}
<?code-excerpt "design_bad.dart (non-inferred-type-args)"?>
{% prettify dart tag=pre+code %}
var playerScores = {};
final events = StreamController();
{% endprettify %}

Sometimes the invocation occurs as the initializer to a variable declaration. If

> variable is *not* a local then instead of writing the type argument list on the

Reviewer: ```suggestion
the variable is *not* local, then instead of writing the type argument list on the
```

Dart do the work for you.

{:.good}
<?code-excerpt "design_good.dart (redundant)"?>
{% prettify dart tag=pre+code %}
class Downloader {
  final Completer<String> response = Completer();
}
{% endprettify %}

{:.bad}
<?code-excerpt "design_bad.dart (redundant)"?>
{% prettify dart tag=pre+code %}
class Downloader {
  final Completer<String> response = Completer<String>();
}
{% endprettify %}

Here, the type annotation on the field provides a surrounding context to infer
the type argument of constructor call in the initializer.

{:.good}
<?code-excerpt "design_good.dart (explicit)"?>
{% prettify dart tag=pre+code %}
var items = Future.value([1, 2, 3]);
{% endprettify %}

{:.bad}
<?code-excerpt "design_bad.dart (explicit)"?>
{% prettify dart tag=pre+code %}
var items = Future<List<int>>.value(<int>[1, 2, 3]);
{% endprettify %}

Here, the types of the collection and instance can be inferred bottom-up from
their elements and arguments.

### AVOID writing incomplete generic types.

When writing a type annotation or type argument, it's still possible to not pin
down a complete type by writing the name of a generic type but omitting its type

> arguments. In Java, these are called "raw types". For example:

Reviewer: I'm having a hard time parsing the first sentence. Could you please rewrite it?

Author: Done!

{:.bad}
<?code-excerpt "design_bad.dart (incomplete-generic)"?>
{% prettify dart tag=pre+code %}
List numbers = [1, 2, 3];
var completer = Completer<Map>();
{% endprettify %}

Here, `numbers` has a type annotation, but the annotation doesn't provide a type
argument to the generic `List`. Likewise, the `Map` type argument to `Completer`
isn't fully specified. In cases like this, Dart will *not* try to "fill in" the
rest of the type for you using the surrounding context. Instead, it silently
fills in any missing type arguments with with `dynamic` (or the bound if the
class has one). That's rarely what you want.

> Instead, if you are writing a generic type in a type annotation or as a type

Reviewer: This sentence is a bit hard to parse. Here's an attempt at making it a bit clearer grammatically:

```suggestion
Instead, if you're writing a generic type either in a type annotation or as a type
```

argument inside some invocation, make sure to write a complete type:

{:.good}
<?code-excerpt "design_good.dart (incomplete-generic)"?>
{% prettify dart tag=pre+code %}
List<num> numbers = [1, 2, 3];
var completer = Completer<Map<String, int>>();
{% endprettify %}

### DO annotate with `dynamic` instead of letting inference fail.

When inference doesn't fill in a type, it usually defaults to `dynamic`. If
`dynamic` is the type you want, this is technically the most terse way to get
it. However, it's not the most *clear* way. A casual reader of your code who

> sees an annotation is missing has no way of knowing if you intended it to be

Reviewer: ```suggestion
sees that an annotation is missing has no way of knowing if you intended it to be
```

mergeJson(original, changes) => ...
{% endprettify %}

Note that it's OK to omit the type when Dart *successfully* infers `dynamic`.

{:.good}
<?code-excerpt "design_good.dart (infer-dynamic)"?>
{% prettify dart tag=pre+code %}
Map<String, dynamic> readJson() => ...

void printUsers() {
  var json = readJson();
  var users = json['users'];
  print(users);
}
{% endprettify %}

Here, Dart infers `Map<String, dynamic>` for `json` and then from that infers
`dynamic` for `users`. It's fine to leave `users` without a type annotation. The
distinction is a little subtle. It's OK to allow inference to *propagate*
`dynamic` through your code from a `dynamic` type annotation somewhere else, but

> you don't want it to summon `dynamic` *ex nihilo* when nothing in your code ever

Reviewer: I'd never seen *ex nihilo* before, and I've seen a lot of Latin! I'm going to pretend I didn't see it, since it's probably against style but is totally cool (and doesn't really affect the meaning of the sentence).
